src/overlays/actors/ovl_En_Dns/z_en_dns.c

```diff
@@ -21,63 +56,466 @@ const ActorInit En_Dns_InitVars = {
+        return 0;
+    }
+    return 4;
+}
+
+u32 func_809EF854(EnDns* this) {
+    if (CHECK_QUEST_ITEM(QUEST_GORON_RUBY) == 0) {
+        return 3;
+    }
+    if (AMMO(ITEM_BOMB) >= CUR_CAPACITY(UPG_BOMB_BAG)) {
+        return 1;
+    }
+    if (gSaveContext.rupees < this->dnsItemEntry->itemPrice) {
+        return 0;
+    }
+    return 4;
+}
+
+u32 func_809EF8F4(EnDns* this) {
+    if (Item_CheckObtainability(ITEM_BOW) == ITEM_NONE) {
+        return 3;
+    }
+    if (AMMO(ITEM_BOW) >= CUR_CAPACITY(UPG_QUIVER)) {
+        return 1;
+    }
+    if (gSaveContext.rupees < this->dnsItemEntry->itemPrice) {
+        return 0;
+    }
+    return 4;
+}
+
+u32 func_809EF9A4(EnDns* this) {
+    if (Inventory_HasEmptyBottle() == 0) {
+        return 1;
+    }
+    if (gSaveContext.rupees < this->dnsItemEntry->itemPrice) {
+        return 0;
+    }
+    return 4;
+}
+
+/* Paying and flagging functions */
+
+void func_809EF9F8(EnDns* this) {
+    Rupees_ChangeBy(-this->dnsItemEntry->itemPrice);
+}
+
+void func_809EFA28(EnDns* this) {
+    Rupees_ChangeBy(-this->dnsItemEntry->itemPrice);
+}
+
+void func_809EFA58(EnDns* this) {
+    gSaveContext.itemGetInf[0] |= 0x800;
+    Rupees_ChangeBy(-this->dnsItemEntry->itemPrice);
+}
+
+void func_809EFA9C(EnDns* this) {
+    Rupees_ChangeBy(-this->dnsItemEntry->itemPrice);
+}
+
+void func_809EFACC(EnDns* this) {
+    Rupees_ChangeBy(-this->dnsItemEntry->itemPrice);
+}
+
+void func_809EFAFC(EnDns* this) {
+    gSaveContext.infTable[25] |= 0x4;
+    Rupees_ChangeBy(-this->dnsItemEntry->itemPrice);
+}
+
+void func_809EFB40(EnDns* this) {
+    gSaveContext.infTable[25] |= 0x8;
+    Rupees_ChangeBy(-this->dnsItemEntry->itemPrice);
+}
+
+void EnDns_SetupWait(EnDns* this, GlobalContext* globalCtx) {
+    if (this->skelAnime.animCurrentFrame == this->skelAnime.animFrameCount) {
+        this->actionFunc = EnDns_Wait;
+        EnDns_ChangeAnim(this, 0);
+    }
+}
+
+void EnDns_Wait(EnDns* this, GlobalContext* globalCtx) {
+    Math_SmoothScaleMaxMinS(&this->actor.shape.rot.y, this->actor.yawTowardsLink, 3, 2000, 0);
+    this->actor.posRot.rot.y = this->actor.shape.rot.y;
+    if (func_8002F194(&this->actor, globalCtx)) {
+        this->actionFunc = EnDns_Talk;
+        return;
+    }
+    if (((this->collider.base.maskA & 2) != 0) || (this->actor.unk_10C != 0)) {
+        this->actor.flags |= 0x10000;
+    } else {
+        this->actor.flags &= 0xFFFEFFFF;
+    }
+    if (this->actor.xzDistFromLink < 130.0f) {
+        func_8002F2F4(&this->actor, globalCtx);
+    }
+}
+
+void EnDns_Talk(EnDns* this, GlobalContext* globalCtx) {
+    if ((func_8010BDBC(&globalCtx->msgCtx) == 4) && (func_80106BC8(globalCtx))) {
+        switch (globalCtx->msgCtx.choiceIndex) {
+            case 0: // OK
+                switch (this->dnsItemEntry->purchaseableCheck(this)) {
+                    case 0:
+                        func_8010B720(globalCtx, 0x10A5);
+                        // "[sound 3882]Not enough Rupees!  Come back again!"
```

Comment:
the comments on the cases for the choice index are fine, but id remove the text id ones

Reply:
I find it easier to see what NPCs with lots of dialog are doing if I can match it to the text, but I can remove them if you'd prefer

Reply:
yeah we've decided against adding the strings into the code as it can clutter up really fast. when we have the text table in the repo it will be very easy to search